spec.md

```diff
@@ -0,0 +1,127 @@
+# SBD Spec
+
+This SBD spec defines the protocol for SBD servers and clients to communicate.
+SBD is a simple websocket-based message relay protocol. 
+
+## 1. Websocket Stack
+
+The SBD protocol is built upon websockets.
+
+### 1.1. Websocket Configuration
+
+#### 1.1.1. Message and Frame Size
+
+The maximum SBD message size (including 32 byte header) is 20000 bytes.
+
+SBD clients and servers MAY set the max message size in the websocket library to 20000 to help enforce this.
+
+The maximum frame size MUST be set larger than 20000 so that sbd messages always fit in a single websocket frame.
+
+## 2. Cryptography
+
+### 2.1. TLS
+
+Generally over the WAN SBD servers SHOULD be available over TLS (wss://), and on a LAN without (ws://).
+
+### 2.2. Ed25519
+
+Clients will be identified by ed25519 public key. Client sessions will be validated by ed25519 signature.
+
+## 3. Protocol
+
+### 3.1. Connect Path
+
+Clients MUST specify exactly 1 http path item on the websocket connection url.
+This item must be the base64url encoded public key that this client will be identified by.
+This public key SHOULD be unique to this new connection.
+
+### 3.2. Messages
+
+#### 3.2.1. Header
+
+SBD messages always contain first a 32 byte header. Messages less than 32 bytes are invalid.
+
+If the header starts with 28 zero bytes, the message is a "command". Otherwise, the message is a "forward".
```

Comment:
If I did the math right (2^28), that puts the probability of a "forward" being misinterpreted as a "command" at around 1 in 268,000,000. That's just the floor -- the probability goes up if e.g. a proof-of-work scheme is used which requires leading zeroes for pub keys.

What are the implications of this misinterpretation? How broadly will those effects spread across the entire network? If a network has 10k agents, the probability of a bad agent key is 1 in 30k, which is beginning to look risky if the effects are widespread.

Reply:
@maackle - though derived from random seed, ed25519 pub keys are not random numbers, my impression is that it would be even more unlikely to result in 28 initial zeros. But regardless, I can put in a simple check to reject public keys that start with 28 zeroes.

Reply:
Done: https://github.com/holochain/sbd/pull/2/commits/fade996672d496cede71fa835d7130eb94986eba